pkg/core/statesync/module.go

```diff
@@ -429,6 +512,81 @@ func (s *Module) AddMPTNodes(nodes [][]byte) error {
 	return nil
 }
 
+// AddContractStorageData adds a batch of key-value pairs for storage-based sync.
+func (s *Module) AddContractStorageData(kvs []KeyValue, syncHeight uint32, expectedRoot util.Uint256) error {
+	if s.mode == MPTBased {
+		panic("contract storage items are not expected in MPT-based mode")
+	}
+	oldStage := s.syncStage
+	s.lock.Lock()
+	defer func() {
+		if s.syncStage != oldStage {
+			s.notifyStageChanged()
+		}
+	}()
+	defer s.lock.Unlock()
+
+	if s.syncStage&headersSynced == 0 || s.syncStage&mptSynced != 0 || expectedRoot.Equals(s.localTrie.StateRoot()) {
+		return errors.New("contract storage items were not requested")
+	}
+	if syncHeight != s.syncPoint {
+		return fmt.Errorf("invalid sync height: got %d, expected %d", syncHeight, s.syncPoint)
+	}
+
+	var (
+		prefix  = TemporaryPrefix(s.dao.Version.StoragePrefix)
+		lastKey []byte
+		batch   = make(map[string][]byte, len(kvs))
+	)
+	for _, kv := range kvs {
+		s.dao.Store.Put(append([]byte{byte(prefix)}, kv.Key...), kv.Value)
+		batch[string(kv.Key)] = kv.Value
+		lastKey = kv.Key
+	}
+	if _, err := s.dao.Store.PersistSync(); err != nil {
+		return fmt.Errorf("failed to persist contract storage data: %w", err)
+	}
+
+	if len(batch) > 0 {
+		mptBatch := mpt.MapToMPTBatch(batch)
+		if _, err := s.localTrie.PutBatch(mptBatch); err != nil {
+			return fmt.Errorf("failed to apply MPT batch at height %d: %w", syncHeight, err)
+		}
+		s.localTrie.Flush(syncHeight)
+	}
+	ckpt := checkpoint{
+		MPTRoot: s.localTrie.StateRoot(),
+		LastKey: lastKey,
+	}
+
+	bw := io.NewBufBinWriter()
+	ckpt.EncodeBinary(bw.BinWriter)
+	if bw.Err != nil {
+		return fmt.Errorf("failed to encode checkpoint metadata: %w", bw.Err)
+	}
+
+	computedRoot := s.localTrie.StateRoot()
+	if !computedRoot.Equals(expectedRoot) {
+		s.dao.Store.Put([]byte{byte(storage.SYSStateSyncCheckpoint)}, bw.Bytes())
+		if _, err := s.dao.Store.PersistSync(); err != nil {
+			return fmt.Errorf("failed to persist checkpoint metadata: %w", err)
+		}
+		return nil
+	}
+
+	s.dao.Store.Delete([]byte{byte(storage.SYSStateSyncCheckpoint)})
```

Comment:
Even though `s.lastStoredKey` is used once by StateFetcher, we need to update this value somewhere in this function. It may not be needed for the current implementation, but may be needed for future usages of `LastStoredKey()`, otherwise this behaviour may cause bugs in future.